src/org/rascalmpl/uri/classloaders/SourceLocationClassLoader.java

```diff
@@ -94,35 +96,63 @@ private List<ClassLoader> initialize(Iterable<? extends IValue> locs) {
         return result;
     }
     
+    @Override
+    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+        Class<?> cls;
+
+        try {
+            // the delegation contract demands us to look up in the parent first.
+            cls = super.loadClass(name, resolve);
+        } 
+        catch (ClassNotFoundException e) {
+            // only then we try our own loop
+            cls = findClass(name);
+        }
+
+        if (resolve) {
+            // optimizes future lookups
+            resolveClass(cls);
+        }
+
+        return cls;
+    }
+
     /**
      * Is called by loadClass but not before searching in the parent classloader.
      */
     @Override
     protected Class<?> findClass(String name) throws ClassNotFoundException {
-        var cached = cache.computeIfAbsent(name, (n) -> {
-            var fileName = n.replace('.', '/') + ".class";
+        Class<?> cached = cache.get(name);
+        if (cached != null) {
+            return cached;
+        }
+
+        synchronized (getClassLoadingLock(name)) {
+            // someone could have added something to the cache by now.
+            cached = cache.get(name);
+            if (cached != null) {
+                return cached;
+            }
+
+            String fileName = name.replace('.', '/') + ".class";
 
             for (ClassLoader l : path) {
                 try (var stream = l.getResourceAsStream(fileName)) {
                     if (stream == null) {
                         continue;
                     }
-                    return defineClass(n, ByteBuffer.wrap(stream.readAllBytes()), null);
-                }
-                catch (IOException e) {
+
+                    byte[] bytes = stream.readAllBytes();
+                    Class<?> cls = defineClass(name, ByteBuffer.wrap(bytes), null);
+                    cache.put(name, cls);
+                    return cls;
```

Comment:
I don't understand why this is better than the `computeIfAbsent`.

The only problem with computeIfAbsent is that it can race for calculating the value (2 threads both doing a `defineClass` for class `ABC`) but the assigned entry in the map is guaranteed to be the same across threads.